rfc/006-semantic-versioning.md

```diff
@@ -0,0 +1,76 @@
+
+Finally I would like to remark that this RFC only addresses the what, as in what are the rules for versioning. The how, or how does Tendermint plan to follow the versioning logic, will be addressed in a later RFC on Soft Upgrades.
+
+## Discussion
+
+We first begin with a round up of the various users and a set of assumptions on what these users expect from Tendermint in regards to versioning:
+
+1. Application Developers, those that use the ABCI to build applications on top of Tendermint, are chiefly concerned with that API. Breaking changes will force developers to modify large portions of their codebase to accommodate for the changes. Some ABCI changes such as introducing priority for the mempool don't require any effort and can be lazily adopted whilst changes like ABCI++ may force applications to redesign their entire execution system.
+2. Node Operators, those running node infrastructure, are predominantly concerned with downtime, complexity and frequency of upgrading and loss of data. They may be also concerned about changes that may break the scripts and tooling they use to supervise their nodes.
+3. External clients, for example, wallets and block explorers are those that perform any of the following:
+     - consume the RPC endpoints of nodes like `/block`
+     - subscribe to events that are streamed via websockets
+     - make queries to the indexer
+  
+    This set are concerned with chain upgrades which will impact their ability to query state and block data as well as broadcast transactions.
+  
+4. There is also a fourth emerging user: IBC module and relayers (although in many ways relayers can also be considered as external clients). The developers of IBC and consumers of their software are concerned about changes that may affect a chains ability to send arbitrary messages to another chain. Specifically, this can happen with a breaking change to the light client verification algorithm.  
+
+In a more broader sense, these four users are inextricably linked and their concerns reflect that of the end user of the applications. The major conclusion that this RFC makes is that **the ability for chains to provide continual service is more important than the actual upgrade burden put on the developers of these chains**. What that means is that although it may be burdensome for application developers to upgrade Tendermint versions, this is far outweighed by the pain node operators experience in upgrading, by the pain that external clients have in accommodating changes and ultimately the pain of end users who are benefiting from the application. Developers have all the time in world to make these changes whilst for other user groups, upgrading can be considered part of this real-time critical path.  
```

Comment:
Very reasonable conclusion 💯

Reply:
> Developers have all the time in world to make these changes whilst for other user groups, upgrading can be considered part of this real-time critical path.

I'm not totally sure I understand what this means. I think we're considering imposing costs in different areas of the ecosystem and it seems like the notion here is that having developers do extra legwork keeps overall costs of a change much lower by shielding external clients from too much additional work. 

I would add that developers don't exactly have _all the time in the world_ since they are trying to ship features and fix bugs, and costly upgrades keep them from satisfying user needs in other ways. I think what this is getting at is that operators, and clients to an extent, experience upgrades as _downtime_, whereas painful upgrades performed by developers create longer release cycles. One pain is a bit more acute than the other.

Reply:
@williambanfield, yeah I agree _all the time in the world_ is an exaggeration and I should be a little more precise in my wording. I think your description more accurately conveys what I was trying to say.